src/transformStateWithClones.js

```diff
@@ -7,7 +7,40 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let current = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear': {
+        current = {};
+        break;
+      }
+
+      case 'addProperties': {
+        const extra = action.extraData || {};
+
+        current = { ...current, ...extra };
+        break;
+      }
+
+      case 'removeProperties': {
+        const toRemove = new Set(action.keysToRemove || []);
+
+        current = Object.fromEntries(
+          Object.entries(current).filter(([key]) => !toRemove.has(key)),
+        );
+        break;
+      }
+
+      default:
+        break;
+    }
+
+    history.push({ ...current });
+  }
```

Comment:
can we rewrite it and make it shorter and more readable (maybe without using the break statement)?